module/localisation/FieldLocalisationNLopt/src/FieldLocalisationNLopt.cpp

```diff
@@ -0,0 +1,346 @@
+            setup_fieldline_distance_map(fd);
+            // Generate the field landmarks
+            setup_field_landmarks(fd);
+            if (cfg.save_map) {
+                std::ofstream file("recordings/fieldline_map.csv");
+                file << fieldline_distance_map.get_map();
+                file.close();
+            }
+
+            // Set the initial state as either left, right, both sides of the field or manually specified inital state
+            auto left_side =
+                Eigen::Vector3d((fd.dimensions.field_length / 4), (fd.dimensions.field_width / 2), -M_PI_2);
+            auto right_side =
+                Eigen::Vector3d((fd.dimensions.field_length / 4), (-fd.dimensions.field_width / 2), M_PI_2);
+            switch (cfg.starting_side) {
+                case StartingSide::LEFT: cfg.initial_hypotheses.emplace_back(left_side); break;
+                case StartingSide::RIGHT: cfg.initial_hypotheses.emplace_back(right_side); break;
+                case StartingSide::EITHER:
+                    cfg.initial_hypotheses.emplace_back(left_side);
+                    cfg.initial_hypotheses.emplace_back(right_side);
+                    break;
+                case StartingSide::CUSTOM: cfg.initial_hypotheses.emplace_back(cfg.initial_state); break;
+                default: log<NUClear::ERROR>("Invalid starting_side specified"); break;
+            }
+            state = cfg.initial_hypotheses[0];
+
+            main_loop_handle.disable();
+            emit<Scope::DELAY>(std::make_unique<ResetFieldLocalisation>(), cfg.start_time_delay);
+            emit(std::make_unique<Stability>(Stability::UNKNOWN));
+        });
+
+        on<Trigger<ResetFieldLocalisation>>().then([this] {
+            log<NUClear::INFO>("Resetting field localisation");
+            main_loop_handle.enable();
+            state   = cfg.initial_hypotheses[0];
+            startup = true;
+        });
+
+        main_loop_handle = on<Trigger<FieldLines>, With<FieldIntersections>, With<Stability>, With<RawSensors>>().then(
+            "NLopt field localisation",
+            [this](const FieldLines& field_lines,
+                   const FieldIntersections& field_intersections,
+                   const Stability& stability,
+                   const RawSensors& raw_sensors) {
+                // Emit field message using ground truth if available
+                if (cfg.use_ground_truth_localisation) {
+                    auto field(std::make_unique<Field>());
+                    field->Hfw = raw_sensors.localisation_ground_truth.Hfw;
+                    emit(field);
+                    return;
+                }
+                // start timeer
+                auto start = std::chrono::steady_clock::now();
+
+                bool unstable = stability == Stability::FALLEN || stability == Stability::FALLING;
+                if (!unstable && field_lines.rPWw.size() > cfg.min_field_line_points) {
```

Comment:
this is just a stylistic thing I'm a fan of to reduce mental load so up to you if you want to do it:
change this big nested `if` to 
```
// Don't run localisation if unstable or don't have enough field line points
if (unstable || field_lines.rPWw.size() < cfg.min_field_line_points) {
    return;
}
// localisation code
```